{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from BCBio import GFF\n",
    "import operator\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "import glob, os\n",
    "from Bio import SeqIO\n",
    "import time\n",
    "import pysam\n",
    "\n",
    "import gffutils\n",
    "import pyfaidx\n",
    "from pysam import VariantFile\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "sourcebase= '/mnt/test_data/hepavac34/rna_benign/spladdrout'\n",
    "#sourcebase = '/tmp/'\n",
    "reffile = '/mnt/test_data/refs/GRCh37.primary_assembly.genome.fa'\n",
    "fileending = 'gff3'\n",
    "vcffile= '/tmp/vcf.vcf.gz' #must be indexed!\n",
    "\n",
    "#db = gffutils.create_db('/tmp/ensg228794.gff', ':memory:')\n",
    "fasta = pyfaidx.Fasta('/mnt/test_data/refs/GRCh37.primary_assembly.genome.fa')\n",
    "vcffile='/tmp/vcf.vcf.gz'\n",
    "\n",
    "outfile = '/tmp/predicted_exons.fa'\n",
    "outfile2 = '/tmp/predicted_genes.fa'  \n",
    "vcfrecords=VariantFile(vcffile)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Getting all gff-files in base directory\n",
    "def get_gff_files(sourcebase):\n",
    "    print \"Checking for GFF Files in directory...\"\n",
    "    gfffiles = []\n",
    "    os.chdir(sourcebase)\n",
    "    for file in glob.glob(\"*.%s\" %fileending):\n",
    "        gfffiles.append(file)\n",
    "    print \"Found %i file(s) in %s ending with %s.\" %(len(gfffiles), sourcebase, fileending)\n",
    "    return gfffiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_gene_sequences(gene,db):\n",
    "    sequences = []\n",
    "    geneseq = []\n",
    "    num_mod=0\n",
    "    #print gene\n",
    "    for i in db.children(gene,featuretype='exon'):\n",
    "        events=check_feature_for_vcfevent(i)\n",
    "        if len(events) == 0:\n",
    "            rec = i.sequence(fasta)#.translate(to_stop=True)\n",
    "            sequences.append(SeqRecord(Seq(rec),i.id,i.attributes['Parent'][0],\"\"))\n",
    "            geneseq.append(rec)\n",
    "        else:\n",
    "            num_mod+=1\n",
    "        #    sequences.append(Seq.translate(i.sequence(fasta),to_stop=True))\n",
    "            rec = get_modified_sequence(i.sequence(fasta),events,i.start)\n",
    "            sequences.append(SeqRecord(Seq(rec),i.id,i.attributes['Parent'][0],\"\"))\n",
    "            geneseq.append(rec)\n",
    "            #sequences.append(get_modified_sequence(i.sequence(fasta),events,i.start))\n",
    "            #return get_modified_sequence(i.sequence(fasta),events)\n",
    "        #if i.id!='exon_7569':\n",
    "        #    print \"this is the record\"\n",
    "        #    print sequences[-1]\n",
    "        \n",
    "    return sequences,[SeqRecord(Seq(''.join(geneseq)),gene.id,\"\")], num_mod "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_feature_for_vcfevent(feature):\n",
    "    vcfevents = []\n",
    "    if feature.chrom not in vcfrecords.header.contigs:\n",
    "        print \"Chrom not found\"\n",
    "        return vcfevents\n",
    "    for rec in vcfrecords.fetch(feature.chrom,feature.start,feature.stop):\n",
    "        vcfevents.append(rec)\n",
    "       #a=rec\n",
    "    if len(vcfevents) == 0:\n",
    "        return vcfevents\n",
    "        print \"no events in range detected (%s - %i on %i)\"%(feature.chrom,feature.start,feature.stop)\n",
    "    else:\n",
    "        print \"%i events in range detected (%s - %i on %i)\"%(len(vcfevents),feature.chrom,feature.start,feature.stop)\n",
    "        \n",
    "    return vcfevents"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_modified_sequence(in_seq,events,offset):\n",
    "    print \"Personalisizing sequence len:%i with %i events\"%(len(in_seq),len(events))\n",
    "    lengthchange=0\n",
    "    for event in events:\n",
    "        eventpos=event.start-offset\n",
    "        print \"Eventstart %i, offset: %i -> pos:%i\"%(event.start,offset,eventpos)\n",
    "        print \"Char at  %i: %s (region +-3: %s)\" %(eventpos,in_seq[eventpos],in_seq[eventpos-3:eventpos+3])\n",
    "        alleles=event.alleles\n",
    "        l = list(in_seq)\n",
    "        \n",
    "        if is_snp(alleles):\n",
    "          #  print \"Found SNP substituting %s with %s\" %(l[eventpos],alleles[1])\n",
    "            if l[eventpos]!= alleles[0]:\n",
    "                print \"warning SNP missmatch\"\n",
    "            l[eventpos]=alleles[1]\n",
    "        if is_del(alleles):\n",
    "            l[eventpos]=alleles[1] #TODO falsch!\n",
    "            lengthchange-=len(alleles[1])-len(alleles[0])\n",
    "            print \"Found deletion: deleting %s from sequences\"%alleles[0]\n",
    "            print alleles\n",
    "            print event\n",
    "        if is_insert(alleles):\n",
    "            l.insert(eventpos,alleles[1])\n",
    "            lengthchange+=len(alleles[0])-len(alleles[1])\n",
    "            print \"found insert: inserting %s into sequence at pos %i\"%alleles[1],eventpos\n",
    "            print event\n",
    "    new_seq=''.join(l)\n",
    "    if (len(in_seq)-len(new_seq)) != lengthchange:\n",
    "        print \"Warning: length changed %i (should be %i)\" %(len(in_seq)-len(new_seq),lengthchange)\n",
    "    return new_seq "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_snp(e):\n",
    "    if len(e[0]) == 1 and len(e[1]) == 1:\n",
    "        return True\n",
    "    return False\n",
    "\n",
    "def is_insert(e):\n",
    "    if len(e[0])<len(e[1]):\n",
    "        return True\n",
    "    return False\n",
    "\n",
    "def is_del(e):\n",
    "    if len(e[0])>len(e[1]):\n",
    "        return True\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "def run():\n",
    "    gfffiles = get_gff_files(sourcebase)\n",
    "    \n",
    "    for infile in gfffiles:\n",
    "        total_genes=0\n",
    "        total_exons=0\n",
    "        total_genes_wrote=0\n",
    "        total_exons_wrote=0\n",
    "        total_mod=0\n",
    "        db = gffutils.create_db(infile, ':memory:')\n",
    "        for gene in db.features_of_type('gene'):\n",
    "            total_genes+=1\n",
    "            seqs,geneseq,num_mod=get_gene_sequences(gene,db)\n",
    "            total_exons+=len(seqs)\n",
    "            total_mod+=num_mod\n",
    "            #a=seqs\n",
    "           # return seqs\n",
    "            #return geneseq\n",
    "            total_exons_wrote+=write_records(translate_records(seqs),'exons')\n",
    "            total_genes_wrote+=write_records(translate_records(geneseq),'genes')\n",
    "        print \"processed (and wrote): %s with %i(%i) genes and %i(%i) exons (%i modified))\" %(infile,total_genes,total_genes_wrote,total_exons,total_exons_wrote,total_mod)\n",
    "        \n",
    "    #ref_recs = load_reffile(reffile)\n",
    "    \n",
    "     #   records, records_until_stop = get_records(infile,ref_recs)\n",
    "      #  write_records(records,records_until_stop,os.path.basename(infile))\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [],
   "source": [
    "def write_records(records,type):\n",
    "    #translate records\n",
    "    \n",
    "    if type == 'exons':\n",
    "        with open(outfile, \"a\") as out_handle:\n",
    "            #print(records[1])\n",
    "            return SeqIO.write(records, out_handle, \"fasta\")\n",
    "    if type == 'genes':\n",
    "        with open(outfile2, \"a\") as out_handle:\n",
    "            #print(records[1])\n",
    "            return SeqIO.write(records, out_handle, \"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Checking for GFF Files in directory...\n",
      "Found 5 file(s) in /mnt/test_data/hepavac34/rna_benign/spladdrout ending with gff3.\n",
      "Chrom not found\n",
      "Chrom not found\n",
      "Chrom not found\n",
      "1 events in range detected (chr17 - 26873725 on 26874867)\n",
      "Personalisizing sequence len:1143 with 1 events\n",
      "Eventstart 26874689, offset: 26873725 -> pos:964\n",
      "Char at  964: G (region +-3: CACGGT)\n",
      "warning SNP missmatch\n",
      "1 events in range detected (chr4 - 155491571 on 155491855)\n",
      "Personalisizing sequence len:285 with 1 events\n",
      "Eventstart 155491783, offset: 155491571 -> pos:212\n",
      "Char at  212: C (region +-3: GGCCCT)\n",
      "1 events in range detected (chr4 - 155491571 on 155491951)\n",
      "Personalisizing sequence len:381 with 1 events\n",
      "Eventstart 155491783, offset: 155491571 -> pos:212\n",
      "Char at  212: C (region +-3: GGCCCT)\n",
      "1 events in range detected (chr6 - 109767339 on 109767560)\n",
      "Personalisizing sequence len:222 with 1 events\n",
      "Eventstart 109767448, offset: 109767339 -> pos:109\n",
      "Char at  109: T (region +-3: CCCTGA)\n",
      "1 events in range detected (chr6 - 109767339 on 109767975)\n",
      "Personalisizing sequence len:637 with 1 events\n",
      "Eventstart 109767448, offset: 109767339 -> pos:109\n",
      "Char at  109: C (region +-3: GGGCAT)\n",
      "warning SNP missmatch\n",
      "processed (and wrote): merge_graphs_intron_retention_C3.confirmed.gff3 with 5183(15549) genes and 15549(46647) exons (5 modified))\n",
      "1 events in range detected (chr1 - 16959609 on 16959841)\n",
      "Personalisizing sequence len:233 with 1 events\n",
      "Eventstart 16959813, offset: 16959609 -> pos:204\n",
      "Char at  204: A (region +-3: GCCAGG)\n",
      "warning SNP missmatch\n",
      "1 events in range detected (chr1 - 16959590 on 16959841)\n",
      "Personalisizing sequence len:252 with 1 events\n",
      "Eventstart 16959813, offset: 16959590 -> pos:223\n",
      "Char at  223: G (region +-3: GGTGCT)\n",
      "warning SNP missmatch\n",
      "1 events in range detected (chr11 - 65652549 on 65652657)\n",
      "Personalisizing sequence len:109 with 1 events\n",
      "Eventstart 65652599, offset: 65652549 -> pos:50\n",
      "Char at  50: G (region +-3: CAGGAC)\n",
      "warning SNP missmatch\n",
      "1 events in range detected (chr11 - 65652549 on 65652657)\n",
      "Personalisizing sequence len:109 with 1 events\n",
      "Eventstart 65652599, offset: 65652549 -> pos:50\n",
      "Char at  50: G (region +-3: CAGGAC)\n",
      "warning SNP missmatch\n",
      "1 events in range detected (chr20 - 61588793 on 61588932)\n",
      "Personalisizing sequence len:140 with 1 events\n",
      "Eventstart 61588823, offset: 61588793 -> pos:30\n",
      "Char at  30: A (region +-3: GTCAGC)\n",
      "warning SNP missmatch\n",
      "1 events in range detected (chr20 - 61588793 on 61588932)\n",
      "Personalisizing sequence len:140 with 1 events\n",
      "Eventstart 61588823, offset: 61588793 -> pos:30\n",
      "Char at  30: A (region +-3: GTCAGC)\n",
      "warning SNP missmatch\n",
      "1 events in range detected (chr20 - 33328167 on 33331145)\n",
      "Personalisizing sequence len:2979 with 1 events\n",
      "Eventstart 33328421, offset: 33328167 -> pos:254\n",
      "Char at  254: C (region +-3: CCTCCA)\n",
      "warning SNP missmatch\n",
      "1 events in range detected (chr6 - 109767339 on 109767615)\n",
      "Personalisizing sequence len:277 with 1 events\n",
      "Eventstart 109767448, offset: 109767339 -> pos:109\n",
      "Char at  109: A (region +-3: GGCAGA)\n",
      "warning SNP missmatch\n",
      "1 events in range detected (chr6 - 109767339 on 109767615)\n",
      "Personalisizing sequence len:277 with 1 events\n",
      "Eventstart 109767448, offset: 109767339 -> pos:109\n",
      "Char at  109: A (region +-3: GGCAGA)\n",
      "warning SNP missmatch\n",
      "processed (and wrote): merge_graphs_alt_5prime_C3.confirmed.gff3 with 4847(14541) genes and 19388(58164) exons (9 modified))\n",
      "Chrom not found\n",
      "Chrom not found\n",
      "Chrom not found\n",
      "Chrom not found\n",
      "1 events in range detected (chr1 - 169509532 on 169512229)\n",
      "Personalisizing sequence len:2698 with 1 events\n",
      "Eventstart 169510649, offset: 169509532 -> pos:1117\n",
      "Char at  1117: A (region +-3: TAAATC)\n",
      "warning SNP missmatch\n",
      "1 events in range detected (chr1 - 169509532 on 169512352)\n",
      "Personalisizing sequence len:2821 with 1 events\n",
      "Eventstart 169510649, offset: 169509532 -> pos:1117\n",
      "Char at  1117: A (region +-3: AAAAAA)\n",
      "warning SNP missmatch\n",
      "1 events in range detected (chr14 - 105221967 on 105223262)\n",
      "Personalisizing sequence len:1296 with 1 events\n",
      "Eventstart 105221961, offset: 105221967 -> pos:-6\n",
      "Char at  -6: T (region +-3: GCTTGG)\n",
      "Found deletion: deleting CCCAGAGAAGACCAA from sequences\n",
      "('CCCAGAGAAGACCAA', 'C')\n",
      "chr14\t105221962\t.\tCCCAGAGAAGACCAA\tC\t.\tPASS\tANN=C|frameshift_variant&splice_acceptor_variant&splice_region_variant&intron_variant|HIGH|SIVA1|ENSG00000184990|transcript|ENST00000556195|protein_coding|2/4|c.170-4_179delCCAGAGAAGACCAA|p.Glu57fs||170/579|57/192||WARNING_TRANSCRIPT_NO_START_CODON,C|frameshift_variant&splice_acceptor_variant&splice_region_variant&intron_variant|HIGH|SIVA1|ENSG00000184990|transcript|ENST00000329967|protein_coding|2/4|c.119-4_128delCCAGAGAAGACCAA|p.Glu40fs||119/528|40/175||,C|frameshift_variant&splice_acceptor_variant&splice_region_variant&intron_variant|HIGH|SIVA1|ENSG00000184990|transcript|ENST00000553810|protein_coding|2/4|c.119-4_128delCCAGAGAAGACCAA|p.Glu40fs||119/508|40/168||WARNING_TRANSCRIPT_INCOMPLETE,C|frameshift_variant&splice_acceptor_variant&splice_region_variant&intron_variant|HIGH|SIVA1|ENSG00000184990|transcript|ENST00000553819|nonsense_mediated_decay|2/6|c.119-4_128delCCAGAGAAGACCAA|p.Glu40fs||119/510|40/169||,C|splice_acceptor_variant&splice_region_variant&intron_variant&non_coding_transcript_exon_variant|HIGH|SIVA1|ENSG00000184990|transcript|ENST00000535554|retained_intron|2/3|n.174-4_183delCCAGAGAAGACCAA||||||,C|splice_acceptor_variant&splice_region_variant&intron_variant&non_coding_transcript_exon_variant|HIGH|SIVA1|ENSG00000184990|transcript|ENST00000556431|retained_intron|2/2|n.119-4_128delCCAGAGAAGACCAA||||||,C|sequence_feature|MODERATE|SIVA1|ENSG00000184990|cross-link:Glycyl_lysine_isopeptide_(Lys-Gly)_(interchain_with_G-Cter_in_ubiquitin)|ENST00000329967|protein_coding|2/4|c.119-4_128delCCAGAGAAGACCAA||||||,C|intron_variant|MODIFIER|SIVA1|ENSG00000184990|transcript|ENST00000347067|protein_coding|1/2|c.119-978_119-965delCCAGAGAAGACCAA||||||,C|intron_variant|MODIFIER|SIVA1|ENSG00000184990|transcript|ENST00000554013|retained_intron|1/1|n.140-978_140-965delCCAGAGAAGACCAA||||||;IC=0;IHP=5;LOF=(SIVA1|ENSG00000184990|8|0.50);MQ=60;MQ0=0;NT=ref;QSI=80;QSI_NT=80;RC=1;RU=CCAGAGAAGACCAA;SGT=ref->het;SOMATIC;TQSI=1;TQSI_NT=1;rna_tum_freq=0;rna_tum_depth=654;rna_ref_freq=0;rna_ref_depth=474\tBCN50:DP:DP2:DP50:FDP50:SUBDP50:TAR:TIR:TOR\t0:32:32:29.24:0:0:39,39:0,0:0,0\t0:24:24:21.12:0:0:15,15:10,10:2,3\n",
      "\n",
      "Warning: length changed 0 (should be 14)\n",
      "1 events in range detected (chr19 - 49611052 on 49611736)\n",
      "Personalisizing sequence len:685 with 1 events\n",
      "Eventstart 49611467, offset: 49611052 -> pos:415\n",
      "Char at  415: G (region +-3: AGCGAG)\n",
      "warning SNP missmatch\n",
      "1 events in range detected (chr19 - 49611079 on 49611736)\n",
      "Personalisizing sequence len:658 with 1 events\n",
      "Eventstart 49611467, offset: 49611079 -> pos:388\n",
      "Char at  388: G (region +-3: AGCGAG)\n",
      "warning SNP missmatch\n",
      "1 events in range detected (chr6 - 109767339 on 109767560)\n",
      "Personalisizing sequence len:222 with 1 events\n",
      "Eventstart 109767448, offset: 109767339 -> pos:109\n",
      "Char at  109: T (region +-3: CCCTGA)\n",
      "1 events in range detected (chr6 - 109767339 on 109767560)\n",
      "Personalisizing sequence len:222 with 1 events\n",
      "Eventstart 109767448, offset: 109767339 -> pos:109\n",
      "Char at  109: T (region +-3: CCCTGA)\n",
      "1 events in range detected (chr6 - 109767339 on 109767560)\n",
      "Personalisizing sequence len:222 with 1 events\n",
      "Eventstart 109767448, offset: 109767339 -> pos:109\n",
      "Char at  109: T (region +-3: CCCTGA)\n",
      "1 events in range detected (chr6 - 109767339 on 109767615)\n",
      "Personalisizing sequence len:277 with 1 events\n",
      "Eventstart 109767448, offset: 109767339 -> pos:109\n",
      "Char at  109: A (region +-3: GGCAGA)\n",
      "warning SNP missmatch\n",
      "1 events in range detected (chr7 - 100417179 on 100417511)\n",
      "Personalisizing sequence len:333 with 1 events\n",
      "Eventstart 100417336, offset: 100417179 -> pos:157\n",
      "Char at  157: G (region +-3: CGGGGG)\n",
      "warning SNP missmatch\n",
      "1 events in range detected (chr7 - 100417179 on 100417511)\n",
      "Personalisizing sequence len:333 with 1 events\n",
      "Eventstart 100417336, offset: 100417179 -> pos:157\n",
      "Char at  157: G (region +-3: CGGGGG)\n",
      "warning SNP missmatch\n",
      "1 events in range detected (chr8 - 22460050 on 22460099)\n",
      "Personalisizing sequence len:50 with 1 events\n",
      "Eventstart 22460062, offset: 22460050 -> pos:12\n",
      "Char at  12: C (region +-3: TCCCAC)\n",
      "warning SNP missmatch\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "processed (and wrote): merge_graphs_alt_3prime_C3.confirmed.gff3 with 7030(21090) genes and 28120(84360) exons (12 modified))\n",
      "1 events in range detected (chr14 - 105221967 on 105222161)\n",
      "Personalisizing sequence len:195 with 1 events\n",
      "Eventstart 105221961, offset: 105221967 -> pos:-6\n",
      "Char at  -6: A (region +-3: CCGAAG)\n",
      "Found deletion: deleting CCCAGAGAAGACCAA from sequences\n",
      "('CCCAGAGAAGACCAA', 'C')\n",
      "chr14\t105221962\t.\tCCCAGAGAAGACCAA\tC\t.\tPASS\tANN=C|frameshift_variant&splice_acceptor_variant&splice_region_variant&intron_variant|HIGH|SIVA1|ENSG00000184990|transcript|ENST00000556195|protein_coding|2/4|c.170-4_179delCCAGAGAAGACCAA|p.Glu57fs||170/579|57/192||WARNING_TRANSCRIPT_NO_START_CODON,C|frameshift_variant&splice_acceptor_variant&splice_region_variant&intron_variant|HIGH|SIVA1|ENSG00000184990|transcript|ENST00000329967|protein_coding|2/4|c.119-4_128delCCAGAGAAGACCAA|p.Glu40fs||119/528|40/175||,C|frameshift_variant&splice_acceptor_variant&splice_region_variant&intron_variant|HIGH|SIVA1|ENSG00000184990|transcript|ENST00000553810|protein_coding|2/4|c.119-4_128delCCAGAGAAGACCAA|p.Glu40fs||119/508|40/168||WARNING_TRANSCRIPT_INCOMPLETE,C|frameshift_variant&splice_acceptor_variant&splice_region_variant&intron_variant|HIGH|SIVA1|ENSG00000184990|transcript|ENST00000553819|nonsense_mediated_decay|2/6|c.119-4_128delCCAGAGAAGACCAA|p.Glu40fs||119/510|40/169||,C|splice_acceptor_variant&splice_region_variant&intron_variant&non_coding_transcript_exon_variant|HIGH|SIVA1|ENSG00000184990|transcript|ENST00000535554|retained_intron|2/3|n.174-4_183delCCAGAGAAGACCAA||||||,C|splice_acceptor_variant&splice_region_variant&intron_variant&non_coding_transcript_exon_variant|HIGH|SIVA1|ENSG00000184990|transcript|ENST00000556431|retained_intron|2/2|n.119-4_128delCCAGAGAAGACCAA||||||,C|sequence_feature|MODERATE|SIVA1|ENSG00000184990|cross-link:Glycyl_lysine_isopeptide_(Lys-Gly)_(interchain_with_G-Cter_in_ubiquitin)|ENST00000329967|protein_coding|2/4|c.119-4_128delCCAGAGAAGACCAA||||||,C|intron_variant|MODIFIER|SIVA1|ENSG00000184990|transcript|ENST00000347067|protein_coding|1/2|c.119-978_119-965delCCAGAGAAGACCAA||||||,C|intron_variant|MODIFIER|SIVA1|ENSG00000184990|transcript|ENST00000554013|retained_intron|1/1|n.140-978_140-965delCCAGAGAAGACCAA||||||;IC=0;IHP=5;LOF=(SIVA1|ENSG00000184990|8|0.50);MQ=60;MQ0=0;NT=ref;QSI=80;QSI_NT=80;RC=1;RU=CCAGAGAAGACCAA;SGT=ref->het;SOMATIC;TQSI=1;TQSI_NT=1;rna_tum_freq=0;rna_tum_depth=654;rna_ref_freq=0;rna_ref_depth=474\tBCN50:DP:DP2:DP50:FDP50:SUBDP50:TAR:TIR:TOR\t0:32:32:29.24:0:0:39,39:0,0:0,0\t0:24:24:21.12:0:0:15,15:10,10:2,3\n",
      "\n",
      "Warning: length changed 0 (should be 14)\n",
      "1 events in range detected (chr20 - 33328167 on 33331145)\n",
      "Personalisizing sequence len:2979 with 1 events\n",
      "Eventstart 33328421, offset: 33328167 -> pos:254\n",
      "Char at  254: C (region +-3: CCTCCA)\n",
      "warning SNP missmatch\n",
      "1 events in range detected (chr8 - 22460050 on 22460156)\n",
      "Personalisizing sequence len:107 with 1 events\n",
      "Eventstart 22460062, offset: 22460050 -> pos:12\n",
      "Char at  12: C (region +-3: TCCCAC)\n",
      "warning SNP missmatch\n",
      "processed (and wrote): merge_graphs_exon_skip_C3.confirmed.gff3 with 9779(29337) genes and 48895(146685) exons (3 modified))\n",
      "1 events in range detected (chr1 - 16959609 on 16959841)\n",
      "Personalisizing sequence len:233 with 1 events\n",
      "Eventstart 16959813, offset: 16959609 -> pos:204\n",
      "Char at  204: A (region +-3: GCCAGG)\n",
      "warning SNP missmatch\n",
      "1 events in range detected (chr1 - 16959609 on 16959841)\n",
      "Personalisizing sequence len:233 with 1 events\n",
      "Eventstart 16959813, offset: 16959609 -> pos:204\n",
      "Char at  204: A (region +-3: GCCAGG)\n",
      "warning SNP missmatch\n",
      "1 events in range detected (chr11 - 112072746 on 112072913)\n",
      "Personalisizing sequence len:168 with 1 events\n",
      "Eventstart 112072843, offset: 112072746 -> pos:97\n",
      "Char at  97: A (region +-3: GTCAAG)\n",
      "1 events in range detected (chr11 - 112072746 on 112072913)\n",
      "Personalisizing sequence len:168 with 1 events\n",
      "Eventstart 112072843, offset: 112072746 -> pos:97\n",
      "Char at  97: A (region +-3: GTCAAG)\n",
      "processed (and wrote): merge_graphs_mult_exon_skip_C3.confirmed.gff3 with 1777(5331) genes and 11749(35247) exons (4 modified))\n"
     ]
    }
   ],
   "source": [
    "a=run()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "def translate_records(records):\n",
    "    new_records = []\n",
    "    for rec in records:\n",
    "        for i in range(0,3):\n",
    "            new_records.append(SeqRecord(rec.seq[i:].translate(to_stop=True),\"%s %s TL-Windows: %i \"%(rec.name,rec.id,i),rec.name,\"\"))\n",
    "    return new_records"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "b=translate_records(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('AGTGAATATACGTGCCCTGTTTGTTGAATCTACTCATCCTTAAAGGCTGAATGC...AAG', Alphabet()), id='intron_retention.6', name='', description='<unknown description>', dbxrefs=[])"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "write_records(b,'genes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
